### Ensembled_Technique & Random Forest

There are 3 Ensembled Technique 

1.Bagging

2.Boosting

3.Stacking

#### Bagging ~ Bootstrap Aggregation
It Create Random Sample of Datapoint(Random Train Set) with Sampling with Replacement(A Particular Datapoints can be repeated in multiple Random sample)

Independent of Previous Model (Parallel in nature)

prediction Average / mode is taken

In [1]:
# Bagged Decision Trees for Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier

#loading the Dataset
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
Dataset = pd.read_csv('C:/Users/Akaash/Downloads/pima-indians-diabetes_data.csv', names=names)

#Separating the X & Y Variables
array = Dataset.values
X = array[:,0:8]
Y = array[:,8]

# Creating K-Fold Instance
seed = 7
kfold = KFold(n_splits=10,random_state=seed,shuffle = True)

#Creating the Instance of DecisionTreeClassifier()
classifier = DecisionTreeClassifier()

#Training the Model -- Bagging
num_trees = 100
#base_estimator= instance of the Technique used which is either DTclassifier or DTRegressor
model = BaggingClassifier(base_estimator=classifier,n_estimators=num_trees,random_state=seed) 

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)

In [2]:
# Accuracy for Each K-Fold
results

array([0.76623377, 0.75324675, 0.74025974, 0.77922078, 0.80519481,
       0.79220779, 0.66233766, 0.75324675, 0.78947368, 0.73684211])

In [3]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

75.78263841421736

Inference: Using the Bagging Ensembled Technique the Final Accuracy is 75.78%

#### Random Forest
It is Based on the Concept of Bagging where Datapoint are Randomly Sampled.

It Splits Data Point as well as the features Randomly and Creates Decision Tree

In [4]:
# Random Forest Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

#loading the Dataset
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
Dataset = pd.read_csv('C:/Users/Akaash/Downloads/pima-indians-diabetes_data.csv', names=names)

#Separating the X & Y Variables
array = Dataset.values
X = array[:,0:8]
Y = array[:,8]

# Creating K-Fold Instance
kfold = KFold(n_splits=10, random_state=7,shuffle = True)

#Training the Model ~ Random Forest 
num_trees = 100
max_features = 3 
model = RandomForestClassifier(n_estimators=num_trees, max_features=max_features) #max_features will sample randomly 3 Features

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)

In [5]:
# Accuracy for Each K-Fold
results

array([0.80519481, 0.80519481, 0.72727273, 0.77922078, 0.80519481,
       0.79220779, 0.64935065, 0.80519481, 0.77631579, 0.76315789])

In [6]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

77.08304853041696

Inference: Using The Random Forest the Final Accuracy is 76.44%

#### Boosting ~ AdaBoost
Stage Weight of That Datapoint Who give Error is updated in the next Training Set of Next model, and the Next Model Try to minimize the Error / Boost the Performance of the model

It is Dependent on the Previous model i.e sequential in Nature

In [7]:
# AdaBoost Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import AdaBoostClassifier

#loading the Dataset
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
Dataset = pd.read_csv('C:/Users/Akaash/Downloads/pima-indians-diabetes_data.csv', names=names)

#Separating the X & Y Variables
array = Dataset.values
X = array[:,0:8]
Y = array[:,8]

# Creating K-Fold Instance
seed=7
kfold = KFold(n_splits=10, random_state=seed,shuffle = True)

#Training the Model ~ Adaboost (Boosting)
num_trees = 10
model = AdaBoostClassifier(n_estimators=num_trees, random_state=seed)

#Predicting & Getting the Model Score with kfold
results = cross_val_score(model, X, Y, cv=kfold)


In [8]:
# Accuracy for Each K-Fold
results

array([0.80519481, 0.74025974, 0.67532468, 0.81818182, 0.80519481,
       0.79220779, 0.71428571, 0.75324675, 0.78947368, 0.77631579])

In [9]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

76.69685577580314

Inference: Using the Boosting Ensembled Technique the Final Accuracy is 76.69%

#### Stacking

Through Stacking Can Use Multiple Algorithms At Once the TAle mean/Mode of the Prediction

In [10]:
# Stacking Ensemble for Classification

#Importing the Required Libraries
import pandas as pd
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import VotingClassifier

#loading the Dataset
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
Dataset = pd.read_csv('C:/Users/Akaash/Downloads/pima-indians-diabetes_data.csv', names=names)

#Separating the X & Y Variables
array = Dataset.values
X = array[:,0:8]
Y = array[:,8]

# Creating K-Fold Instance
seed=7
kfold = KFold(n_splits=10, random_state=seed,shuffle = True)

# create the sub models - 3 Different Models
estimators = []
model1 = LogisticRegression(max_iter=500)
estimators.append(('logistic', model1))
model2 = DecisionTreeClassifier()
estimators.append(('cart', model2))
model3 = SVC()
estimators.append(('svm', model3))

# create the ensemble model
ensemble = VotingClassifier(estimators)

#Predicting & Getting the Model Score with kfold
results = cross_val_score(ensemble, X, Y, cv=kfold)


In [11]:
# Accuracy for Each K-Fold
results

array([0.83116883, 0.75324675, 0.74025974, 0.80519481, 0.79220779,
       0.79220779, 0.63636364, 0.79220779, 0.81578947, 0.76315789])

In [12]:
#Average of All k-fold Accuracy to get Final model Accuary
results.mean()*100.0

77.21804511278197

Inference: Using the Stacking Ensembled Technique the Final Accuracy is 77.21%